src/Nethermind/Nethermind.Blockchain/OnChainRemoveTxWatcher.cs

```diff
@@ -0,0 +1,72 @@
+//  Copyright (c) 2021 Demerzel Solutions Limited
+//  This file is part of the Nethermind library.
+// 
+//  The Nethermind library is free software: you can redistribute it and/or modify
+//  it under the terms of the GNU Lesser General Public License as published by
+//  the Free Software Foundation, either version 3 of the License, or
+//  (at your option) any later version.
+// 
+//  The Nethermind library is distributed in the hope that it will be useful,
+//  but WITHOUT ANY WARRANTY; without even the implied warranty of
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+//  GNU Lesser General Public License for more details.
+// 
+//  You should have received a copy of the GNU Lesser General Public License
+//  along with the Nethermind. If not, see <http://www.gnu.org/licenses/>.
+// 
+
+using System;
+using System.Threading.Tasks;
+using Nethermind.Blockchain.Receipts;
+using Nethermind.Core;
+using Nethermind.Logging;
+
+namespace Nethermind.Blockchain
+{
+    public class OnChainRemoveTxWatcher : IDisposable
+    {
+        private readonly IBlockTree _blockTree;
+        private readonly IReceiptStorage _receiptStorage;
+        private readonly ILogger _logger;
+
+        public OnChainRemoveTxWatcher(IBlockTree blockTree, IReceiptStorage receiptStorage, ILogManager logManager)
```

Comment:
use nullable reference types in any new code that you write

Reply:
the naming is really wrong here:
it is not chain removal
it is not tx watcher

you can maybe call it ReceiptCanonicalityMonitor / ReceiptCanonicalityManager

Reply:
It's ReceiptCanonicalityMonitor now